modules/noise/noise.h

```diff
@@ -0,0 +1,239 @@
+			}
+		}
+
+		// Blend the vertical skirt over the middle seam.
+		for (int x = half_width; x < skirt_edge_x; x++) {
+			int alpha = 255 * (1 - Math::smoothstep(.1f, .9f, float(x - half_width) / float(skirt_width)));
+			for (int y = 0; y < p_height; y++) {
+				// Skip the center square
+				if (y == half_height) {
+					y = skirt_edge_y - 1;
+				} else {
+					// Starts reading at s2, ALT_Y skips s3, and continues with s1.
+					wr(x, y) = _alpha_blend<T>(rd_dest(x, y), rd_src(x, y, img_buff<T>::ALT_Y), alpha);
+				}
+			}
+		}
+
+		// Blend the horizontal skirt over the middle seam.
+		for (int y = half_height; y < skirt_edge_y; y++) {
+			int alpha = 255 * (1 - Math::smoothstep(.1f, .9f, float(y - half_height) / float(skirt_height)));
+			for (int x = 0; x < p_width; x++) {
+				// Skip the center square
+				if (x == half_width) {
+					x = skirt_edge_x - 1;
+				} else {
+					// Starts reading at s4, skips s3, continues with s5.
+					wr(x, y) = _alpha_blend<T>(rd_dest(x, y), rd_src(x, y, img_buff<T>::ALT_X), alpha);
+				}
+			}
+		}
+
+		// Fill in the center square. Wr starts at the top left of Q4, which is the equivalent of the top left of s3, unless a modulo is used.
+		for (int y = half_height; y < skirt_edge_y; y++) {
+			for (int x = half_width; x < skirt_edge_x; x++) {
+				int xpos = 255 * (1 - Math::smoothstep(.1f, .9f, float(x - half_width) / float(skirt_width)));
+				int ypos = 255 * (1 - Math::smoothstep(.1f, .9f, float(y - half_height) / float(skirt_height)));
+
+				// Blend s3(Q1) onto s5(Q2) for the top half.
+				T top_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_X), rd_src(x, y, img_buff<T>::DEFAULT), xpos);
+				// Blend s1(Q3) onto Q4 for the bottom half.
+				T bottom_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_XY), rd_src(x, y, img_buff<T>::ALT_Y), xpos);
+				// Blend the top half onto the bottom half.
+				wr(x, y) = _alpha_blend<T>(bottom_blend, top_blend, ypos);
+			}
+		}
+		Ref<Image> image = memnew(Image(p_width, p_height, false, p_src->get_format(), dest));
+		p_src.unref();
+		return image;
+	}
+
+	template <typename T>
+	T _alpha_blend(T p_bg, T p_fg, int p_alpha) const {
+		l2c fg, bg, out;
+
+		fg.l = p_fg;
+		bg.l = p_bg;
+
+		uint16_t alpha;
+		uint16_t inv_alpha;
+
+		// If no alpha argument specified, use the alpha channel in the color
+		if (p_alpha == -1) {
+			alpha = fg.c[3] + 1;
+			inv_alpha = 256 - fg.c[3];
+		} else {
+			alpha = p_alpha + 1;
+			inv_alpha = 256 - p_alpha;
+		}
+
+		out.c[0] = (uint8_t)((alpha * fg.c[0] + inv_alpha * bg.c[0]) >> 8);
+		out.c[1] = (uint8_t)((alpha * fg.c[1] + inv_alpha * bg.c[1]) >> 8);
+		out.c[2] = (uint8_t)((alpha * fg.c[2] + inv_alpha * bg.c[2]) >> 8);
+		out.c[3] = 0xFF;
+
+		return out.l;
+	}
+
+protected:
+	static void _bind_methods();
+
+public:
+	// Virtual destructor so we can delete any Noise derived object when referenced as a Noise*.
+	virtual ~Noise() {}
+
+	virtual real_t get_noise_1d(real_t p_x) = 0;
```

Comment:
Does pure virtual means extensions cannot define new noises?

Reply:
`virtual` means referencing a `Noise*` that points to a `FastNoiseLite` object will call the FastNoiseLite implementation in the derived class rather than the implementation in `Noise` without casting.

`= 0` means there is no implementation, thus one cannot instantiate Noise().

I don't know about GDExtensions, but when I wrote this as a module this was so I could define OpenSimplexNoise, FastNoise, and FastNoiseSIMD modules.

Reply:
I just noticed that the functions should actually bound with GDVIRTUAL_BIND(), but I am not exacly sure.